### PR TITLE
Feature refactor create servergroup

### PR DESCRIPTION
**Pull request by drmaas**
Refactoring create and clone server groups to use auto scaling groups and template parameters instead of raw templates.

### CODE REVIEW

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy [hunk 1]
```diff
@@ -43,44 +42,62 @@ class CloneOpenstackAtomicOperation implements AtomicOperation<DeploymentResult>
...
+      securityGroups = description.serverGroupParameters?.securityGroups ?: ancestorParams.securityGroups
+    }
+    deployDescription.disableRollback = description.disableRollback ?: false
+    deployDescription.timeoutMins = description.timeoutMins ?: ancestorStack.timeoutMins
+    deployDescription.region = description.region
+
+    task.updateStatus BASE_PHASE, "Finished reating new server group description"
 
-    return newDescription
+    return deployDescription
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy]: I think the standard has been not to use the return keyword so far.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy [hunk 2]
```diff
@@ -41,20 +45,48 @@ class DeployOpenstackAtomicOperation implements AtomicOperation<Void> {
...
 
-    task.updateStatus BASE_PHASE, "Looking up next sequence index for cluster ${groupName}..."
-    def stackName = serverGroupNameResolver.resolveNextServerGroupName(description.application, description.stack, description.freeFormDetails, false)
-    task.updateStatus BASE_PHASE, "Heat stack name chosen to be ${stackName}."
+      task.updateStatus BASE_PHASE, "Looking up next sequence index for cluster ${groupName}..."
+      def stackName = serverGroupNameResolver.resolveNextServerGroupName(description.application, description.stack, description.freeFormDetails, false)
+      task.updateStatus BASE_PHASE, "Heat stack name chosen to be ${stackName}."
 
+      task.updateStatus BASE_PHASE, "Loading templates"
+      String templateFile = 'asg.yaml'
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy]: Should we move these (yaml file names) to a constants file?

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/domain/ServerGroupParameters.groovy [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.openstack.domain
+
+import groovy.transform.AutoClone
+import groovy.transform.Canonical
+
+/**
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/domain/ServerGroupParameters.groovy:22]: Remove empty comments.

File: clouddriver-openstack/src/main/resources/asg.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      max_size: {get_param: max_size}
+      resource:
+        type: asg_resource.yaml
+        properties:
+          flavor: {get_param: flavor}
+          image: {get_param: image}
+          internal_port: {get_param: internal_port}
+          network_id: {get_param: network_id}
+          pool_id: {get_param: pool_id}
+          security_groups: {get_param: security_groups}
```
Comment (derekolk) [inline clouddriver-openstack/src/main/resources/asg.yaml:44]: Don't we need to add scaling policies to add/remove instances or is that defined in other operations?  Thought we would have to create default settings if not specified.

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidatorSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+import com.netflix.spinnaker.clouddriver.openstack.deploy.description.servergroup.DeployOpenstackAtomicOperationDescription
+import com.netflix.spinnaker.clouddriver.openstack.domain.ServerGroupParameters
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackCredentials
+import com.netflix.spinnaker.clouddriver.openstack.security.OpenstackNamedAccountCredentials
+import com.netflix.spinnaker.clouddriver.security.AccountCredentialsProvider
+import org.springframework.validation.Errors
+import spock.lang.Specification
+
+/**
```
Comment (derekolk) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidatorSpec.groovy]: Remove empty comment block.

File: clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidatorSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+    when:
+    validator.validate([], description, errors)
+
+    then:
+    0 * errors.rejectValue(_,_)
+  }
+
+  def "Validate application - error"() {
+    given:
```
Comment (derekolk) [inline clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/servergroup/DeployOpenstackAtomicOperationValidatorSpec.groovy]: Might be able to dry this up using this strategy ... https://github.com/spinnaker/clouddriver/blob/master/clouddriver-openstack/src/test/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/loadbalancer/UpsertOpenstackLoadBalancerDescriptionValidatorSpec.groovy#L116

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy [hunk 7]
```diff
@@ -41,20 +45,48 @@ class DeployOpenstackAtomicOperation implements AtomicOperation<Void> {
...
+      String template = IOUtils.toString(this.class.classLoader.getResourceAsStream(templateFile))
+      String subtemplateFile = 'asg_resource.yaml'
+      String subtemplate = IOUtils.toString(this.class.classLoader.getResourceAsStream(subtemplateFile))
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy]: Feels like IOUtils could be extended to load the resource and take in multiple templates return a map of resources.

File: clouddriver-openstack/src/main/resources/asg.yaml (hunk 4 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/resources/asg.yaml:44]: We do. I think it should be a separate feature.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy (hunk 1 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/CloneOpenstackAtomicOperation.groovy]: Wasn't my code :) will remove.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy (hunk 2 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy]: No one else will need access. So I will just make constant in this file.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/domain/ServerGroupParameters.groovy (hunk 3 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/domain/ServerGroupParameters.groovy:22]: Added actual comment.

File: clouddriver-openstack/src/main/resources/asg.yaml (hunk 4 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/resources/asg.yaml:44]: I wanted to get it working first. I'll try adding in non-parameterized defaults for now.

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy (hunk 7 above)
Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy]: Template and subtemplate need to be passed to openstack4j as separate args, not sure what this gains us.

Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/servergroup/DeployOpenstackAtomicOperation.groovy]: The gain is not repeating ourselves, lines 62 - 65 look like copy/paste. 

Thinking something like this if you don't want to modify IOUtils:
`
Map<String, String> templateMap = ['asg.yaml', 'asg_resource.yaml'].collectEntries {
   [(it): IOUtils.toString(this.class.classLoader.getResourceAsString(it))]
}
`

File: clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/OpenstackAttributeValidator.groovy [hunk 8]
```diff
@@ -109,9 +109,9 @@ class OpenstackAttributeValidator {
     result
   }
 
-  boolean validateNonNegative(int value, String attribute) {
+  boolean validateNonNegative(Integer value, String attribute) {
     def result
-    if (value >= 0) {
+    if (value != null && value >= 0) {
```
Comment (derekolk) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/OpenstackAttributeValidator.groovy:114]: Use groovy truthiness.

Comment (drmaas) [inline clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/validators/OpenstackAttributeValidator.groovy:114]: Can't, 0 evaluates to null

**Comment by derekolk** [comment]
👍

**Pull request closed without merging**